src/mono/browser/runtime/cancelable-promise.ts

```diff
@@ -32,17 +38,121 @@ export function wrap_as_cancelable<T>(inner: Promise<T>): ControllablePromise<T>
 export function mono_wasm_cancel_promise(task_holder_gc_handle: GCHandle): void {
     const holder = _lookup_js_owned_object(task_holder_gc_handle) as PromiseHolder;
     mono_assert(!!holder, () => `Expected Promise for GCHandle ${task_holder_gc_handle}`);
+    holder.cancel();
+}
 
-    const promise = holder.promise;
-    loaderHelpers.assertIsControllablePromise(promise);
-    const promise_control = loaderHelpers.getPromiseController(promise);
-    if (holder.isResolved) {
-        // FIXME: something needs to free the GCHandle
-        mono_log_warn("Canceling a promise that has already resolved.");
-        return;
-    }
-    mono_assert(!holder.isCanceled, "This promise already canceled.");
-    holder.isCanceled = true;
-    promise_control.reject(new Error("OperationCanceledException"));
+// NOTE: layout has to match PromiseHolderState in JSHostImplementation.Types.cs
+const enum PromiseHolderState {
+    IsResolving = 0,
 }
 
+const promise_holder_symbol = Symbol.for("wasm promise_holder");
+
+export class PromiseHolder extends ManagedObject {
+    public isResolved = false;
```

Comment:
i feel like if a key part of this object's state lives in the wasm heap, we should put all of the state there and access it using atomic/volatile APIs, instead of splitting some of the state into JS and some into native. so i'd like to see isResolved and isPosted move into the native heap. not mandatory though, it just is harder to reason about how this code works with the state split like it is right now

Reply:
this is here just for sanity sake, I will change it into assert